### PR TITLE
Update to support simulation of different client's speed in async mode

### PR DESCRIPTION
**Pull request by cuiboyuan**
## Description

In the async mode, most of the clients have a relatively fast speed, so it is sometimes quite hard to test Plato in a scenario where clients have various different speeds. This change allows users to simulate clients' speeds by providing a distribution in the configuration file. The user can also choose to only enable the simulation without providing a specific distribution, and the code would just use a default one.

Currently, the simulation only supports Zipf and Normal distribution. More distributions can be added in the future.

## How has this been tested?

- Test 1 with default distribution:
 Run `./run -c ./configs/MNIST/fedavg_async_lenet5.yml`
- Test 2 with Normal distribution: 
 First update the client configuration in **configs/MNIST/fedavg_async_lenet5.yml** as below:
```
clients:
    type: simple

    total_clients: 2

    per_round: 2

    do_test: false

    simulation: true

    simulation_distribution:
        distribution: normal
        mean: 2
        sd: 1
```
 Then run  `./run -c ./configs/MNIST/fedavg_async_lenet5.yml`
- Test 3 with Zipf distribution:
 First update the client configuration in **configs/MNIST/fedavg_async_lenet5.yml** as below:
```
clients:
    type: simple

    total_clients: 2

    per_round: 2

    do_test: false

    simulation: true

    simulation_distribution:
        distribution: zipf
        s: 2
```
 Then run  `./run -c ./configs/MNIST/fedavg_async_lenet5.yml`

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue) Fixes #
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

## Additional information
The current implementation deviates a bit from the original intent. The original intents were to put clients to sleep at the end of each epoch. To accomplish that, the `await asyncio.sleep()` code should be inserted in method `Trainer.train_process()` in file **plato/trainers/basic.py**, and that would require the method `Trainer.train_process()` to be changed to `async`, and we need to change every function that calls `train_process()` to `async`. I'm afraid that might break the code, so I just stay with the current implementation, where the clients are put to sleep after they finished the model training.

UPDATE: please ignore the information above, the code is now implemented in a way such that clients are put to sleep at the end of each epoch. Please refer to the conversation below for details.

### CODE REVIEW

**Comment by baochunli** [comment]
We may still need to sleep at the end of each epoch, rather than at the end of each communication round. The training loop will need to be modified anyway due to the need for training to terminate prematurely when the servers requests a client to do so, so implementing a finer granularity sleep simulation in the training process may not be so risky and detrimental.

**Comment by cuiboyuan** [comment]
I see. That makes a lot of sense. The code is now modified such that clients are put to sleep at the end of each epoch in the simulation.

Thanks.

**Comment by baochunli** [comment]
It appears that when `max_concurrency` in the trainer is not set, the clients will use exactly the same sleep time. Perhaps this is because the current code uses `mp.Manager().dict()` to store the sleep times simulated, yet without `max_concurrency` set, no processes will be spawned.

**Comment by cuiboyuan** [comment]
It seems like the problem is related to the random seed. I tried to run `fedavg_async_lenet5.yml` with `random_seed` and `max_concurrency` not set, and the clients had different sleep times.

I guess the seed is set sometime before `train_process()` is called, but with `max_concurrency` set, function `train_process()` is executed in a new process, so the memory related to the seed does not pass along. However with`max_concurrecy` not set, no new process will be spawned, so the random seed will cause the clients to have exactly the same sleep time.

Thanks.

**Comment by cuiboyuan** [comment]
I have updated the implementation by removing the use of `mp.Manager().dict()` as it is causing memory leaks on my end (running on Windows 10) as well as failing the auto-tests.

Thanks.

File: plato/trainers/basic.py [hunk 1]
```diff
@@ -447,3 +509,15 @@ async def server_test(self, testset, sampler=None):
...
+        """
+            Obtain a saved model for a particular epoch that finishes just after the provided
+            wall clock time is reached.
+        """
+        print("len of models per epoch = ", len(self.models_per_epoch))
+        print(self.models_per_epoch)
+        for epoch in sorted(self.models_per_epoch):
+            model = self.models_per_epoch[epoch]
+            if model['time'] + self.tic > wall_time:
+                return model['model']
```
Comment (cuiboyuan) [inline plato/trainers/basic.py:523]: Just some questions about the urgent message implementations. Why do we need to find the model per epoch by the wall clock time? I think here we want to find the model that is trained by the most recent epoch, so could we just return the model with the largest epoch number in `self.models_per_epoch`?

Thanks!

Comment (baochunli) [reply plato/trainers/basic.py:523]: The wall clock time was previously provided by the server, and that was the simulated wall clock time when the previous client (before the one that the urgent request needs to be sent to) had finished training. This is when the server needs to traverse all the training clients and send them the urgent request to stop training prematurely.

However, the current implementation does have a bug: it assumes that `train_process()` is sharing memory with its parent process, which is not the case. You can see from debugging output that the length of self.models_per_epoch is zero. This is incorrect.

To fix this, we will (unfortunately) have to "remember" these models by writing to checkpoints, and then reading from one these checkpoints when `obtain_model_update()` is called.

Comment (cuiboyuan) [reply plato/trainers/basic.py]: I see. That makes sense. Thanks!

**Comment by cuiboyuan** [comment]
I have fixed a bug related to `obtain_model_update()` in `trainers/basic.py`. The trainer originally uses `self.tic + training_time` to compare with `wall_time`. However, `self.tic` is initialized by `time.perf_counter()`, while `wall_time` is initialized by `time.time()` on the server side. These two numbers differ significantly (at least on my side), so the function never really finds and returns the model closest to `wall_time` with checkpoints files.

I updated the code such that `obtain_model_update()` uses a new variable `self.training_start_time` to replace `self.tic`, and `self.training_start_time` is set to `time.time()` at the beginning of function `train()`. This way, the function actually returns the model right after the time when sever requests for urgent model update.

**Comment by baochunli** [comment]
Makes a lot of sense! Thank you! There could be other issues in the current codebase; I'm still testing it.

**Comment by cuiboyuan** [comment]
No problem. I will also continue to test the code.

Thanks

**Comment by cuiboyuan** [comment]
I have added the implementation for FedAsync. The overall structure is completed, but currently there is a small bug where an AssertionError will be raised from `base.Server.choose_clients()` if the total number of clients is the same as the number of clients selected per round, and I'm still working on resolving it.

Thanks.

**Comment by baochunli** [comment]
I revised the design of FedAsync, as the previous design was faulty as it bypassed all the existing mechanisms on simulating the wall clock time and asynchronous mode.

**Comment by cuiboyuan** [comment]
I see. The revised structure makes more sense. But one small issue is that since we are now using `process_client()` of class `base.Server`, the client technically does not get processed immediately after it reports back in FedAsync. The client reports will need to wait for at most `periodic_interval` seconds to be picked up and processed by `periodic_task()`. Would that be an issue? Or is it a delay we can tolerate for FedAsync?

Thanks

**Comment by baochunli** [comment]
No, this is not an issue as long as the periodic interval is sufficiently short, and much shorter than the typical time needed for local training on a client.

**Comment by cuiboyuan** [comment]
Ok I see. Thanks!

**Comment by baochunli** [comment]
> No, this is not an issue as long as the periodic interval is sufficiently short, and much shorter than the typical time needed for local training on a client.

Also, the duration of the periodic interval does not lead to any inaccuracy to the experimental results, since we will be simulating the wall clock time, taking advantage of all the recent work towards this direction.

**Comment by cuiboyuan** [comment]
Thanks! That makes a lot of sense.

I have also taken a deeper look at the bug of AssertionError of `select_clients()`. It seems like this is not specific to FedAsync, and it is not really a bug. Generally for asynchronous mode (without setting `simulation`, `simulate_wall_clock`, and `request_update`), the total number of clients (`total_clients`) needs to be at least twice the number of clients selected per round (`per_round`) to ensure that the AssertionError will not be triggered. Because the reported clients and clients that are still training cannot be re-selected when `select_clients()` is called, there need to be `per_round` idle clients available for selection.

I guess a question would be why couldn't we re-select the clients that already report back (i.e. in the `self.reporting_clients` list) for the next training round?

Thanks

**Comment by baochunli** [comment]
> Thanks! That makes a lot of sense.
> 
> I have also taken a deeper look at the bug of AssertionError of `select_clients()`. It seems like this is not specific to FedAsync, and it is not really a bug. Generally for asynchronous mode (without setting `simulation`, `simulate_wall_clock`, and `request_update`), the total number of clients (`total_clients`) needs to be at least twice the number of clients selected per round (`per_round`) to ensure that the AssertionError will not be triggered. Because the reported clients and clients that are still training cannot be re-selected when `select_clients()` is called, there need to be `per_round` idle clients available for selection.
> 
> I guess a question would be why couldn't we re-select the clients that have already reported back (i.e. in the `self.reporting_clients` list) for the next training round in `select_clients()`?
> 
> Thanks

This is due to the fact that, in asynchronous mode when the wall clock time is simulated, the server will need to wait for all the clients per round to report before it can take action. After it has processed the first `minimum_clients_aggregated` (say, 2) clients, the simulated wall clock time is advanced so that these 2 clients are in the history, but the other clients are still in the future in simulated wall clock time, even though they have already reported. In other words, the wall clock time that's simulated on the server decides what the current time is. The server certainly couldn't select a client that is currently still training locally in real-time, so it couldn't do so in simulated time either.
